Module authorizationserver

PatrykMilewski opened the pull request:
This closes #50

PatrykMilewski committed:
Added Authorization Server logic as Manager Module

Added http client implementation based on Apache Http

Added wrapers for session token, credentials and secret

Added Reader implementation for reading data stored in storage

Took 11 hours 56 minutes

PatrykMilewski committed:
Fixed all problmes, fully working version

There are still missing some method implementations of abstract Module class

Took 1 hour 36 minutes

PatrykMilewski committed:
Merge remote-tracking branch 'origin/master' into module-authorizationserver

# Conflicts:
#	Manager/pom.xml
#	Manager/src/main/java/io/raspberrywallet/manager/cryptography/crypto/exceptions/DecryptionException.java
#	Manager/src/main/java/io/raspberrywallet/manager/modules/ExampleModule.java
#	Manager/src/main/java/io/raspberrywallet/manager/modules/Module.java

Took 8 minutes

PatrykMilewski committed:
Merged with current master

Changed storing secret logic for authorization server module, now Server is storing encryption key, that is used for key part encryption input from manager

Not working version, needs some debugging

Took 1 hour 25 minutes

PatrykMilewski committed:
The code is passing tests, functionally working

Took 35 minutes

stanbar commented on Manager/src/main/java/io/raspberrywallet/manager/common/http/ApacheSecuredHttpClient.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.io.IOException;
+
+public class ApacheSecuredHttpClient extends HttpClient {
```
What is this class for ?

PatrykMilewski replied:
It's for the future, right now auth server is using http and when https will be implemented, then I will change this to correct client

stanbar replied:
Ok

stanbar commented on Manager/src/main/java/io/raspberrywallet/manager/common/http/ApacheHttpClient.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ApacheHttpClient extends HttpClient {
```
I don't get the naming idea, `ApacheHttpClient` extends `HttpClient`, where  `HttpClient` uses `org.apache.` libraries so it's already coupled with Apache, and cannot be reused in other ways.

PatrykMilewski replied:
Isn't it good practice to add another level of abstraction? There will be only single point to change, when we would like to change apache to something else, since we use it through this class

stanbar replied:
Sure, but then the most abstract class (HttpClient) should be independent from Apache. ApacheHttpClient should overlay this dependency

PatrykMilewski replied:
Okay, I understand, but still can't see how to make it right, could you please give me an idea, how should I name these 3 classes?

stanbar replied:
Sure, first of all I would ask a question if we are planing to use more than one httpclients, 
- if yes, then, I would move `JSONObject convert(Form body)` to some Utils class, because it's stateless and make this class dependent of `org.apache.http.client.fluent.Form`.
Then I would generify the class to:
```java
public abstract class HttpClient<Response, Body> {
    public abstract Response sendPOSTRequest(Body body, String endpoint) throws IOException;
    public abstract Response sendGETRequest(Body body, String endpoint);
}
```
- If not I would just rename `HttpClient` to `ApacheHttpClient`, `ApacheHttpClient` to `UnsecureHttpClient` and leave `SecureHttpClient` as is.

And after all, I would rebase onto http://square.github.io/okhttp/  😄

stanbar commented on Manager/src/main/java/io/raspberrywallet/manager/common/http/ApacheHttpClient.java:
```diff
@@ -0,0 +1,58 @@
+    
+    @Override
+    public HttpResponse sendPOSTRequest(Form body, String endpoint) throws IOException {
+        org.apache.http.client.HttpClient client = HttpClients.createDefault();
```
It would be nice if you could reuse OkHttpClient with Jackson, it's already in our dependencies, and seems to be much smaller and have nicer API

stanbar commented on Manager/src/main/java/io/raspberrywallet/manager/common/readers/StringReader.java:
```diff
@@ -0,0 +1,56 @@
+
+import java.security.SecureRandom;
+
+public class StringReader extends Reader<String> {
```
What is this class for ? And why is it called StringReader while it lazly write and returns random string ?

PatrykMilewski replied:
Well it is no longer used, I've forgot to delete this. Will change

stanbar commented on Manager/src/main/java/io/raspberrywallet/manager/common/readers/WalletUUIDReader.java:
```diff
@@ -0,0 +1,46 @@
+import java.nio.file.StandardOpenOption;
+import java.util.UUID;
+
+public class WalletUUIDReader extends Reader<UUID> {
```
What is the Wallet UUID ?

PatrykMilewski replied:
I had to choose idntifier for every wallet on server side. I've decided to use UUID, that is random for every wallet. The secret on the server side is stored as key-value pair, where key is wallet UUID.

stanbar replied:
Ok, looks good

stanbar commented on Manager/src/main/java/io/raspberrywallet/manager/modules/Module.java:
```diff
@@ -81,7 +82,7 @@ public String getStatusString() {
      * @param status - new status
      */
-    void setStatusString(@NotNull String status) {
+    protected void setStatusString(@NotNull String status) {
```
this methond will be used by manager to change the module status. So public modifier is required

PatrykMilewski replied:
will change

stanbar commented on Manager/src/main/java/io/raspberrywallet/manager/modules/authorizationserver/AuthorizationServerModule.java:
```diff
@@ -0,0 +1,137 @@
+    @Nullable
+    @Override
+    public String getHtmlUi() {
+        return "<input type=\"text\" name=\"password\">";
```
so the login is the WalletUUID right ?

PatrykMilewski replied:
yup, wallet manages it by itself, when there is no saved uuid for wallet, then a new one is generated and saved to file as plain text

stanbar replied:
Looks good :)

PatrykMilewski committed:
Renamed HttpClient classes to ApacheHttpClient

Removed unused StringReader class

Removed unused host parameter from HttpClients constructors

Changed access modifiers from protected to public in Module class

Took 42 minutes

PatrykMilewski commented:
@stasbar please review again, I made changes that you asked for, except for okhttp clients, due to some problems, Apache is fine for now, will refactor it sometime to okhttp

stanbar commented on Manager/src/main/java/io/raspberrywallet/manager/common/http/SecuredApacheApacheHttpClient.java:
```diff
@@ -6,7 +6,7 @@
 import java.io.IOException;
 
-public class ApacheSecuredHttpClient extends HttpClient {
+public class SecuredApacheApacheHttpClient extends ApacheHttpClient {
```
Shouldn't it be named just "SecureApacheHttpClient" ?

stanbar commented on Manager/src/main/java/io/raspberrywallet/manager/common/http/UnsecuredApacheApacheHttpClient.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class UnsecuredApacheApacheHttpClient extends ApacheHttpClient {
```
Shouldn't it be named just "UnsecureApacheHttpClient" ?

PatrykMilewski committed:
Fixed class names

Took 24 minutes

PatrykMilewski committed:
Merge with actual master

Took 4 minutes
